// fabric-network/lib/impl/event/contractlistenersession.js
"use strict";
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractListenerSession = void 0;
const Logger = __importStar(require("../../logger"));
const Listeners = __importStar(require("./listeners"));
const logger = Logger.getLogger('ContractListenerSession');
class ContractListenerSession {
    constructor(listener, chaincodeId, network, options) {
        this.listener = listener;
        this.chaincodeId = chaincodeId;
        this.network = network;
        this.blockListener = this.newBlockListener(options);
        this.options = options;
    }
    async start() {
        await this.network.addBlockListener(this.blockListener, this.options);
    }
    close() {
        this.network.removeBlockListener(this.blockListener);
    }
    newBlockListener(options) {
        const callback = this.onContractEvent.bind(this);
        return Listeners.blockFromContractListener(callback, options === null || options === void 0 ? void 0 : options.checkpointer);
    }
    async onContractEvent(event) {
        if (this.isMatch(event)) {
            await this.notifyListener(event);
        }
    }
    isMatch(event) {
        return event.chaincodeId === this.chaincodeId;
    }
    async notifyListener(event) {
        try {
            await this.listener(event);
        }
        catch (error) {
            logger.warn('Error notifying contract listener', error);
        }
    }
}
exports.ContractListenerSession = ContractListenerSession;
