// fabric-network/lib/impl/event/listeners.js
"use strict";
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockFromContractListener = exports.checkpointBlockListener = void 0;
const GatewayUtils = __importStar(require("../gatewayutils"));
const Logger = __importStar(require("../../logger"));
const logger = Logger.getLogger('Listener');
function checkpointBlockListener(listener, checkpointer) {
    return async (blockEvent) => {
        const checkpointBlockNumber = await checkpointer.getBlockNumber();
        if (!checkpointBlockNumber || checkpointBlockNumber.equals(blockEvent.blockNumber)) {
            await listener(blockEvent);
            const nextBlockNumber = blockEvent.blockNumber.add(1);
            await checkpointer.setBlockNumber(nextBlockNumber);
        }
    };
}
exports.checkpointBlockListener = checkpointBlockListener;
function blockFromContractListener(listener, checkpointer) {
    if (checkpointer) {
        const transactionListener = transactionFromContractListener(listener);
        const checkpointTxListener = checkpointTransactionListener(transactionListener, checkpointer);
        const blockListener = blockFromTransactionListener(checkpointTxListener);
        return checkpointBlockListener(blockListener, checkpointer);
    }
    else {
        const transactionListener = transactionFromContractListener(listener);
        return blockFromTransactionListener(transactionListener);
    }
}
exports.blockFromContractListener = blockFromContractListener;
function transactionFromContractListener(listener) {
    return async (transactionEvent) => {
        for (const contractEvent of transactionEvent.getContractEvents()) {
            await listener(contractEvent);
        }
    };
}
function checkpointTransactionListener(listener, checkpointer) {
    return async (transactionEvent) => {
        const checkpointTransactionIds = await checkpointer.getTransactionIds();
        if (!checkpointTransactionIds.has(transactionEvent.transactionId)) {
            await listener(transactionEvent);
            await checkpointer.addTransactionId(transactionEvent.transactionId);
        }
    };
}
function blockFromTransactionListener(listener) {
    return async (blockEvent) => {
        const transactionPromises = blockEvent.getTransactionEvents()
            .filter((transactionEvent) => transactionEvent.isValid)
            .map((transactionEvent) => listener(transactionEvent));
        // Don't use Promise.all() as it returns early if any promises are rejected
        const results = await GatewayUtils.allSettled(transactionPromises);
        logAndThrowErrors(results);
    };
}
function logAndThrowErrors(results) {
    const errors = results
        .filter((result) => result.status === 'rejected')
        .map((result) => result.reason);
    if (errors.length > 0) {
        errors.forEach((error) => logger.warn('Error notifying transaction listener', error));
        throw new Error(`Error notifying listener: ${errors[0].stack || errors[0].message}`);
    }
}
