// fabric-network/lib/impl/event/orderedblockqueue.js
"use strict";
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderedBlockQueue = void 0;
const Long = __importStar(require("long"));
class OrderedBlockQueue {
    constructor(startBlock) {
        this.queue = new Map();
        this.nextBlockNumber = startBlock;
    }
    addBlock(event) {
        const blockNumber = event.blockNumber;
        if (!this.isNewBlockNumber(blockNumber)) {
            return;
        }
        const key = this.blockNumberToKey(blockNumber);
        this.queue.set(key, event);
        if (!this.nextBlockNumber) {
            this.nextBlockNumber = blockNumber;
        }
    }
    getNextBlock() {
        if (!this.nextBlockNumber) {
            return;
        }
        const key = this.blockNumberToKey(this.nextBlockNumber);
        const event = this.queue.get(key);
        if (event) {
            this.queue.delete(key);
            this.nextBlockNumber = this.nextBlockNumber.add(Long.ONE);
        }
        return event;
    }
    getNextBlockNumber() {
        return this.nextBlockNumber;
    }
    size() {
        return this.queue.size;
    }
    isNewBlockNumber(blockNumber) {
        return !this.nextBlockNumber || this.nextBlockNumber.lessThanOrEqual(blockNumber);
    }
    blockNumberToKey(blockNumber) {
        return blockNumber.toString();
    }
}
exports.OrderedBlockQueue = OrderedBlockQueue;
